File: client/python/mesh/prsim/endpoint.py
#
# Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
# TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
#
#
from abc import ABC, abstractmethod
from typing import Generic

from mesh.macro import mpi, spi, T, A


@spi("mesh")
class Endpoint(ABC):
    """
    Like subscriber but synchronized and has return value. For any extension.
    """

    @abstractmethod
    @mpi("${mesh.uname}")
    def fuzzy(self, buff: bytes) -> bytes:
        """
        Invoke endpoint.
        :param buff:
        :return:
        """
        pass


@spi("mesh")
class EndpointSticker(ABC, Generic[T, A]):
    """
    Like subscriber but synchronized and has return value. For any extension.
    """

    @abstractmethod
    def stick(self, varg: T) -> A:
        """
        Invoke endpoint.
        :param varg:
        :return:
        """
        pass

File: client/java/src/main/java/io/be/mesh/prsim/Endpoint.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package io.be.mesh.prsim;

import io.be.mesh.macro.MPI;
import io.be.mesh.macro.SPI;

import java.nio.ByteBuffer;

/**
 * @author coyzeng@gmail.com
 */
@SPI("mesh")
public interface Endpoint {

    /**
     * Fuzzy call with generic param
     * In multi returns, it's an array.
     */
    @MPI("${mesh.uname}")
    ByteBuffer fuzzy(ByteBuffer buff);

    /**
     * @param <I> is the input
     * @param <O> is the output
     */
    interface Sticker<I, O> {

        // Stick with generic param
        O stick(I varg);

    }

}
